File: development/NET/src/server/Da/ServerDaTransaction.cs
using System;
using System.Collections;
using System.Runtime.InteropServices;
using Softing.OPCToolbox.OTB;

namespace Softing.OPCToolbox.Server
{
	/// <summary>
	/// Transaction class: Used for managing the transaction collection requests
	/// </summary>
	/// <include
	///  file='TBNS.doc.xml'
	///  path='//class[@name="DaTransaction"]/doc/*'
	/// />
	public class DaTransaction
	{
		#region //	Constructors
		//----------------------

		/// <summary>
		/// Default Constructor.
		/// </summary>
		/// <param name="aTransactionType">Client wants to read or to write</param>
		/// <param name="aRequestList">Array of requests contained in the transaction</param>
		/// <param name="aSessionHandle">The session's key</param>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  ctor[@name="DaTransaction"]/doc/*'
		/// />
		public DaTransaction(
			EnumTransactionType aTransactionType,
			DaRequest[] aRequestList,
			uint aSessionHandle)
		{
			this.m_type = aTransactionType;
			this.m_requestList.AddRange(aRequestList);

			this.m_sessionHandle = aSessionHandle;
			lock (s_keyBuilderJanitor)
			{
				this.m_key = s_keyBuilder++;
			}

			foreach (DaRequest request in aRequestList)
			{
				request.TransactionKey = this.m_key;
			} //	end foreach
		} //	end public constructor


		//-
		#endregion

		#region	//	Public Virtual Methods
		//--------------------------------

		/// <summary>
		/// Called if a read transaction occurs
		/// </summary>
		/// <returns>
		/// EnumResultCode.E_NOTIMPL. Must be overridden
		///	</returns>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  method[@name="HandleReadRequests"]/doc/*'
		/// />
		public virtual int HandleReadRequests()
		{
			return (int) EnumResultCode.E_NOTIMPL;
		} //	end HandleReadRequests


		/// <summary>
		/// Called if a write transaction occurs
		/// </summary>
		/// <returns>
		/// EnumResultCode.E_NOTIMPL. Must be overridden
		///	</returns>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  method[@name="HandleWriteRequests"]/doc/*'
		/// />
		public virtual int HandleWriteRequests()
		{
			return (int) EnumResultCode.E_NOTIMPL;
		} //	end HandleWriteRequests


		//-
		#endregion

		#region	//	Protected Attributes
		//----------------------------------

		/// <summary>
		/// incremented with each transaction created
		/// </summary>
		private static uint s_keyBuilder = 1;

		/// <summary>
		/// the key builder janotir
		/// </summary>
		private static object s_keyBuilderJanitor = new object();

		/// <summary>
		/// Requests contained in this transaction
		/// </summary>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  attribute[@name="m_requestList"]/doc/*'
		/// />
		protected ArrayList m_requestList = new ArrayList();

		/// <summary>
		/// requests list janitor
		/// </summary>
		private object m_requestListJanitor = new object();

		/// <summary>
		/// Determines whether this is a read or a write Transaction
		/// </summary>
		private readonly EnumTransactionType m_type = EnumTransactionType.READ;

		/// <summary>
		///	the session's Handle
		/// </summary>
		private readonly uint m_sessionHandle = 0;

		/// <summary>
		/// the transaction unique key
		/// </summary>
		private readonly uint m_key = 0;

		//-
		#endregion

		#region //	Public Properties
		//---------------------------

		/// <summary>
		/// the transaction associated Key
		/// </summary>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  property[@name="Key"]/doc/*'
		/// />
		public uint Key
		{
			get { return m_key; } //	end get
		} //	end Key


		/// <summary>
		/// Determines whether this is a read or a write Transaction
		/// </summary>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  property[@name="Type"]/doc/*'
		/// />
		public EnumTransactionType Type
		{
			get { return m_type; } //	end get
		} //	end Type


		/// <summary>
		/// Requests contained in this transaction
		/// </summary>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  property[@name="Requests"]/doc/*'
		/// />
		public ArrayList Requests
		{
			get { return m_requestList; } //	end get
		} //	end Requests


		/// <summary>
		/// Is transaction empty?
		/// </summary>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  property[@name="IsEmpty"]/doc/*'
		/// />
		public bool IsEmpty
		{
			get
			{
				lock (m_requestListJanitor)
				{
					return (m_requestList.Count == 0);
				} //	end lock
			} //	end get
		} //	end IsEmpty


		/// <summary>
		/// Retrieves the session object owning this transaction
		/// </summary>
		/// <returns>
		/// null - Some error occurred
		/// DaSession the session that generated this transaction
		/// </returns>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  property[@name="Session"]/doc/*'
		/// />
		public DaSession Session
		{
			get
			{
				if (m_sessionHandle != 0)
				{
					return (DaSession) Application.Instance.GetSession(m_sessionHandle);
				} //	end if
				else
				{
					return null;
				} //	end if ... else
			} //	end get
		} //	end Session


		//-
		#endregion

		#region //	Public Methods
		//------------------------

		/// <summary>
		/// Completes all requests must be called while or after "handling" requests
		/// </summary>
		/// <returns>
		/// S_OK - Everything was OK
		/// S_FALSE - Everything was OK
		/// The Result should be checked with ResultCode.SUCCEEDED
		/// or with ResultCode.FAILED
		/// </returns>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  method[@name="CompleteRequests"]/doc/*'
		/// />
		public int CompleteRequests()
		{
			int result = (int) EnumResultCode.E_FAIL;

			int count = 0;
			OTValueData[] values = null;
			OTSRequestData[] requests = null;
			int[] results = null;

			try
			{
				lock (m_requestListJanitor)
				{
					count = m_requestList.Count;

					if (count == 0)
					{
						return (int) EnumResultCode.S_FALSE;
					} //	end if

					values = new OTValueData[count];
					requests = new OTSRequestData[count];
					results = new int[count];

					for (int i = 0; i < count; i++)
					{
						DaRequest request = m_requestList[i] as DaRequest;
						if (request != null)
						{
							requests[i].m_sessionHandle = request.SessionHandle;
							requests[i].m_propertyID = request.PropertyId;
							requests[i].m_requestHandle = request.RequestHandle;

							if (request.AddressSpaceElement != null)
							{
								requests[i].m_object.m_userData = (uint) request.AddressSpaceElement.UserData;
								requests[i].m_object.m_objectHandle = request.AddressSpaceElement.ObjectHandle;
							}
							else
							{
								requests[i].m_object.m_userData = 0;
								requests[i].m_object.m_objectHandle = 0;
							} //	end if ... else

							values[i] = new OTValueData();
							if (request.Value != null)
							{
								values[i].m_timestamp = new OTDateTime(request.Value.TimeStamp);
								values[i].m_quality = (ushort) request.Value.Quality;
								values[i].m_value = Marshal.AllocCoTaskMem(ValueQT.VARIANT_SIZE);
								Marshal.GetNativeVariantForObject(request.Value.Data, values[i].m_value);

								results[i] = (int) request.Result;
							}
							else
							{
								values[i].m_quality = (ushort) EnumQuality.BAD;
								values[i].m_value = Marshal.AllocCoTaskMem(ValueQT.VARIANT_SIZE);
								Marshal.GetNativeVariantForObject(null, values[i].m_value);
								results[i] = (int) EnumResultCode.E_UNEXPECTED;
							} //	end if ... else
						}
						else
						{
							requests[i].m_sessionHandle = 0;
							requests[i].m_propertyID = 0;
							requests[i].m_requestHandle = 0;
							requests[i].m_object.m_userData = 0;
							requests[i].m_object.m_objectHandle = 0;

							values[i] = new OTValueData();
							values[i].m_quality = (ushort) EnumQuality.BAD;
							values[i].m_value = Marshal.AllocCoTaskMem(ValueQT.VARIANT_SIZE);
							Marshal.GetNativeVariantForObject(null, values[i].m_value);
							results[i] = (int) EnumResultCode.E_UNEXPECTED;
						} //	end if ... else

						request.RequestState = EnumRequestState.COMPLETED;
					} //	end for

					m_requestList.Clear();
				} //	end lock

				result = OTBFunctions.OTSCompleteRequests(count, requests, results, values);
			}
			catch (Exception e)
			{
				Application.Instance.Trace(
					EnumTraceLevel.ERR, EnumTraceGroup.OPCSERVER,
					"DaTransaction.CompleteRequests", "Exception caught: " + e.ToString());

				result = (int) EnumResultCode.E_FAIL;
			}
			finally
			{
				Application.Instance.ReleaseTransaction(m_key);

				for (int i = 0; i < count; i++)
				{
					OTBFunctions.OTVariantClear(values[i].m_value);
					Marshal.FreeCoTaskMem(values[i].m_value);
				} //	end for
			} //	end try ... catch ... finally

			return result;
		} //	end CompleteRequests


		/// <summary>
		/// Marks the request identified by "requestHandle" completed, as it was internally handled (~cancelled) inside Toolkit Core
		/// </summary>
		/// <returns>
		/// S_OK - Request was found and marked completed
		/// E_NOTFOUND - Request was not found
		/// The Result should be checked with ResultCode.SUCCEEDED
		/// or with ResultCode.FAILED
		/// </returns>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  method[@name="requestCompletedInternally"]/doc/*'
		/// />
		public int findRequestByHandle(uint requestHandle, out DaRequest request)
		{
			request = null;
			DaRequest req;
			lock (m_requestListJanitor)
			{
				int count = m_requestList.Count;
				for (int i = 0; i < count; ++i)
				{
					req = m_requestList[i] as DaRequest;
					if ((req != null) && (req.RequestHandle == requestHandle))
					{
						request = req;
						return (int) EnumResultCode.S_OK;
					}
				}
			} //	end lock
			return (int) EnumResultCode.E_NOTFOUND;
		}

		/// <summary>
		/// Completes a single request and removes it from the transaction
		/// </summary>
		/// <returns>
		/// S_OK - Everything was OK
		/// S_FALSE - Everything was OK
		/// The Result should be checked with ResultCode.SUCCEEDED
		/// or with ResultCode.FAILED
		/// </returns>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  method[@name="CompleteRequest"]/doc/*'
		/// />
		public int CompleteRequest(DaRequest aRequest)
		{
			int result = (int) EnumResultCode.E_FAIL;

			OTValueData[] values = new OTValueData[1];
			OTSRequestData[] requests = new OTSRequestData[1];
			int[] results = new int[1];

			try
			{
				DaRequest request = aRequest as DaRequest;

				if (request != null)
				{
					request.RequestState = EnumRequestState.COMPLETED;

					//	remove the request from the list
					RemoveRequest(request);

					requests[0].m_sessionHandle = request.SessionHandle;
					requests[0].m_propertyID = request.PropertyId;
					requests[0].m_requestHandle = request.RequestHandle;

					if (request.AddressSpaceElement != null)
					{
						requests[0].m_object.m_userData = (uint) request.AddressSpaceElement.UserData;
						requests[0].m_object.m_objectHandle = request.AddressSpaceElement.ObjectHandle;
					}
					else
					{
						requests[0].m_object.m_userData = 0;
						requests[0].m_object.m_objectHandle = 0;
					} //	end if ... else

					values[0] = new OTValueData();
					if (request.Value != null)
					{
						values[0].m_timestamp = new OTDateTime(request.Value.TimeStamp);
						values[0].m_quality = (ushort) request.Value.Quality;
						values[0].m_value = Marshal.AllocCoTaskMem(ValueQT.VARIANT_SIZE);
						Marshal.GetNativeVariantForObject(request.Value.Data, values[0].m_value);

						results[0] = (int) request.Result;
					}
					else
					{
						values[0].m_quality = (ushort) EnumQuality.BAD;
						values[0].m_value = Marshal.AllocCoTaskMem(ValueQT.VARIANT_SIZE);
						Marshal.GetNativeVariantForObject(null, values[0].m_value);
						results[0] = (int) EnumResultCode.E_UNEXPECTED;
					} //	end if ... else
				}
				else
				{
					requests[0].m_sessionHandle = 0;
					requests[0].m_propertyID = 0;
					requests[0].m_requestHandle = 0;
					requests[0].m_object.m_userData = 0;
					requests[0].m_object.m_objectHandle = 0;

					values[0] = new OTValueData();
					values[0].m_quality = (ushort) EnumQuality.BAD;
					values[0].m_value = Marshal.AllocCoTaskMem(ValueQT.VARIANT_SIZE);
					Marshal.GetNativeVariantForObject(null, values[0].m_value);
					results[0] = (int) EnumResultCode.E_UNEXPECTED;
				} //	end if ... else

				result = OTBFunctions.OTSCompleteRequests(1, requests, results, values);
			}
			catch (Exception e)
			{
				Application.Instance.Trace(
					EnumTraceLevel.ERR, EnumTraceGroup.OPCSERVER,
					"DaTransaction.CompleteRequest", "Exception caught: " + e.ToString());

				result = (int) EnumResultCode.E_FAIL;
			}
			finally
			{
				if (this.IsEmpty)
				{
					Application.Instance.ReleaseTransaction(m_key);
				}
				OTBFunctions.OTVariantClear(values[0].m_value);
				Marshal.FreeCoTaskMem(values[0].m_value);
			} //	end try ... catch ... finally

			return result;
		} //	end CompleteRequest


		/// <summary>
		/// Changes the cache value for the Address space elements provided
		/// </summary>
		/// <returns>
		/// E_INVALIDARG - Invalid valueQT was passed
		/// S_OK - Value changed
		/// OTS_E_EXCEPTION - Unexpected error occurred
		/// The Result should be checked with ResultCode.SUCCEEDED
		/// or with ResultCode.FAILED
		/// </returns>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  method[@name="ValuesChanged"]/doc/*'
		/// />
		public int ValuesChanged()
		{
			int result;
			int count = 0;
			OTValueData[] values = null;
			uint[] handles = null;

			try
			{
				lock (m_requestListJanitor)
				{
					count = m_requestList.Count;

					values = new OTValueData[count];
					handles = new uint[count];

					for (int i = 0; i < count; i++)
					{
						OTValueData valueData = new OTValueData();
						DaRequest request = m_requestList[i] as DaRequest;

						if (request != null)
						{
							valueData.m_quality = (ushort) request.Value.Quality;
							valueData.m_timestamp = new OTDateTime(request.Value.TimeStamp);
							valueData.m_value = Marshal.AllocCoTaskMem(ValueQT.VARIANT_SIZE);

							Marshal.GetNativeVariantForObject(request.Value.Data, valueData.m_value);
							handles[i] = request.AddressSpaceElement.ObjectHandle;

							request.Result = EnumResultCode.S_OK;
						}
						else
						{
							valueData.m_quality = (ushort) EnumQuality.BAD;
							valueData.m_value = Marshal.AllocCoTaskMem(ValueQT.VARIANT_SIZE);
							Marshal.GetNativeVariantForObject(null, valueData.m_value);
							handles[i] = 0;
						} //	end if ... else

						values[i] = valueData;
					} //	end for
					result = OTBFunctions.OTSValuesChanged(count, handles, values);
				} //	end lock
			}
			catch (Exception e)
			{
				Application.Instance.Trace(
					EnumTraceLevel.ERR, EnumTraceGroup.OPCSERVER,
					"DaTransaction.ValuesChanged", "Exception caught: " + e.ToString());
				result = (int) EnumResultCode.E_FAIL;
			}
			finally
			{
				for (int i = 0; i < count; i++)
				{
					OTBFunctions.OTVariantClear(values[i].m_value);
					Marshal.FreeCoTaskMem(values[i].m_value);
				} //	end for
			} //	end try ... catch ... finally

			return result;
		} //	end ValuesChanged


		/// <summary>
		/// removes a specific request from the internal requests list
		/// </summary>
		/// <param name="aRequest">
		/// the request to be removed from the transactrion
		/// </param>
		/// <include
		///  file='TBNS.doc.xml'
		///  path='//class[@name="DaTransaction"]/
		///  method[@name="RemoveRequest"]/doc/*'
		/// />
		public void RemoveRequest(DaRequest aRequest)
		{
			if (aRequest == null)
			{
				return;
			} //	end if

			lock (m_requestListJanitor)
			{
				m_requestList.Remove(aRequest);
			} //	end lock
		} //	end removeRequest


		//-
		#endregion
	} //	end class Transaction
} //	end namespace Softing.OPCToolbox.Server

File: development/C++/src/server/Da/ServerDaTransaction.cpp
#include "OSCompat.h"
#include "ServerDaTransaction.h"

#include "ServerDaAddressSpaceElement.h"
#include "ServerDaRequest.h"
#include "ServerDaSession.h"

#include "../ServerApplication.h"
#include "OTServer.h"



using namespace SoftingOPCToolboxServer;

unsigned long DaTransaction::KeyBuilder = 1;
Mutex DaTransaction::KeyBuilderJanitor = Mutex();

//-----------------------------------------------------------------------------
// Constructor
//
DaTransaction::DaTransaction(
	EnumTransactionType aTransactionType,
	std::vector<DaRequest*>& aRequestList,
	unsigned long aSessionHandle):
	m_type(aTransactionType),
	m_sessionHandle(aSessionHandle),
	m_requestList(aRequestList)
{
	KeyBuilderJanitor.lock();
	m_key = KeyBuilder++;
	KeyBuilderJanitor.unlock();

	unsigned long count = (unsigned long)aRequestList.size();

	for (unsigned long i = 0; i < count; i++)
	{
		aRequestList[i]->setTransactionKey(m_key);
	}   //  end for
}   //  end constructor


//-----------------------------------------------------------------------------
// Destructor
//
DaTransaction::~DaTransaction()
{
	m_requestListJanitor.lock();
	unsigned long count = (unsigned long)m_requestList.size();
	DaRequest* current = NULL;

	for (unsigned long i = 0; i < count; i++)
	{
		current = m_requestList[i];

		if (current != NULL)
		{
			delete current;
		}   //  end if
	}   //  end for

	m_requestListJanitor.unlock();
}   //  end destructor


//----------------------------------------------------------------------------
// Function to get the client from which  the Transaction came from
//
//	returns - null Some error occurred
//	returns - a Session referring to the client the transaction came from
//
DaSession* DaTransaction::getSession(void)
{
	if (m_sessionHandle != 0)
	{
		return Application::Instance()->getSession(m_sessionHandle);
	}   //  end if
	else
	{
		return NULL;
	}   //  end if ... else
}   //  end getSession


//----------------------------------------------------------------------------
// Completes all requests must be called while or after "handling" requests
//
// S_OK - Everything was OK
// S_FALSE - not everything ok, but usable results
// The Result should be checked with ResultCode.SUCCEEDED
// or with ResultCode.FAILED
//
long DaTransaction::completeRequests(void)
{
	long result = E_FAIL;
	m_requestListJanitor.lock();
	size_t count = m_requestList.size();

	if (count == 0)
	{
		m_requestListJanitor.unlock();
		return EnumResultCode_S_FALSE;
	}   //  end if

	OTValueData* values = new OTValueData[count];
	OTSRequestData* requests = new OTSRequestData[count];
	long* results = new long[count];

	for (size_t i = 0; i < count; i++)
	{
		if (m_requestList[i] != NULL)
		{
			requests[i].m_clientHandle  = m_requestList[i]->getSessionHandle();
			requests[i].m_propertyID    = m_requestList[i]->getPropertyId();
			requests[i].m_requestHandle = m_requestList[i]->getRequestHandle();
			DaAddressSpaceElement* element = m_requestList[i]->getAddressSpaceElement();

			if (element != NULL)
			{
				requests[i].m_object.m_userData     = element->getUserData();
				requests[i].m_object.m_objectHandle = element->getObjectHandle();
			}
			else
			{
				requests[i].m_object.m_userData     = 0;
				requests[i].m_object.m_objectHandle = 0;
			}   //  end if ... else

			//  fill in the result
			ValueQT* value = m_requestList[i]->getValue();

			if (value != NULL)
			{
				values[i].m_timestamp   = value->getTimeStamp().get();
				values[i].m_quality     = (unsigned char)value->getQuality();
				values[i].m_pValue      = new VARIANT();
				value->getData().copyTo(values[i].m_pValue);
				results[i] = m_requestList[i]->getResult();
			}
			else
			{
				values[i].m_quality     = EnumQuality_BAD;
				values[i].m_pValue      = new VARIANT();
				values[i].m_pValue->vt  = VT_EMPTY;
				results[i]              = EnumResultCode_E_UNEXPECTED;
			}   //  end if ... else
		}
		else
		{
			requests[i].m_clientHandle          = 0;
			requests[i].m_propertyID            = 0;
			requests[i].m_requestHandle         = 0;
			requests[i].m_object.m_userData     = 0;
			requests[i].m_object.m_objectHandle = 0;
			values[i].m_quality     = EnumQuality_BAD;
			values[i].m_pValue      = new VARIANT();
			values[i].m_pValue->vt  = VT_EMPTY;
			results[i] = EnumResultCode_E_UNEXPECTED;
		}   //  end if ... else

		m_requestList[i]->setRequestState(EnumRequestState_COMPLETED);
	}   //  end for

	// empty request list
	for (size_t k = 0; k < m_requestList.size(); k++)
	{
		if (m_requestList[k] != NULL)
		{
			delete m_requestList[k];
			m_requestList[k] = NULL;
		}   //  end if
	}   //  end for

	m_requestList.clear();

	m_requestListJanitor.unlock();

	result = OTSCompleteRequests((long)count, requests, results, values);

	for (size_t j = 0; j < count; j++)
	{
		if (values[j].m_pValue != NULL)
		{
			Variant::variantClear(values[j].m_pValue);
			delete values[j].m_pValue;
			values[j].m_pValue = NULL;
		}   //  end if
	}   //  end for

	//  release the allocated memory
	if (values != NULL)
	{
		delete[] values;
	}   //  end if

	if (results != NULL)
	{
		delete[] results;
	}   //  end if

	if (requests != NULL)
	{
		delete[] requests;
	}   //  end if

	Application::Instance()->releaseTransaction(m_key);
	return result;
}   //  end completeRequests


//----------------------------------------------------------------------------
// completeRequest
//
long DaTransaction::completeRequest(DaRequest* aRequest)
{
	long result = E_FAIL;
	OTSRequestData requestData;
	memset(&requestData, 0, sizeof(OTSRequestData));
	OTValueData valueData;
	memset(&valueData, 0, sizeof(OTValueData));
	long resultData = 0;
	requestData.m_clientHandle  = aRequest->getSessionHandle();
	requestData.m_propertyID    = aRequest->getPropertyId();
	requestData.m_requestHandle = aRequest->getRequestHandle();
	DaAddressSpaceElement* element = aRequest->getAddressSpaceElement();

	if (element != NULL)
	{
		requestData.m_object.m_userData     = element->getUserData();
		requestData.m_object.m_objectHandle = element->getObjectHandle();
	}
	else
	{
		requestData.m_object.m_userData     = 0;
		requestData.m_object.m_objectHandle = 0;
	}   //  end if ... else

	//  fill in the result
	resultData = aRequest->getResult();
	ValueQT* value = aRequest->getValue();

	if (value != NULL)
	{
		valueData.m_timestamp   = value->getTimeStamp().get();
		valueData.m_quality     = (unsigned char)value->getQuality();
		valueData.m_pValue      = new VARIANT();
		value->getData().copyTo(valueData.m_pValue);
	}
	else
	{
		valueData.m_quality     = EnumQuality_BAD;
		valueData.m_pValue      = new VARIANT();
		valueData.m_pValue->vt  = VT_EMPTY;
		resultData              = EnumResultCode_E_UNEXPECTED;
	}   //  end if ... else

	aRequest->setRequestState(EnumRequestState_COMPLETED);
	removeRequest(aRequest);
	result = OTSCompleteRequests(1, &requestData, &resultData, &valueData);

	if (valueData.m_pValue != NULL)
	{
		Variant::variantClear(valueData.m_pValue);
		delete valueData.m_pValue;
		valueData.m_pValue = NULL;
	}   //  end if

	if (this->isEmpty())
	{
		Application::Instance()->releaseTransaction(m_key);
	}   //  end if

	return result;
}   //  end completeRequest

//-----------------------------------------------------------------------------
// requestCompletedInternally
//
long DaTransaction::findRequestByHandle(unsigned long requestHandle, DaRequest** aRequest)
{
	long ret = OT_E_OPC_NOTFOUND;
	(*aRequest) = NULL;
	std::vector<DaRequest*>::iterator requestIterator;
	m_requestListJanitor.lock();
	for (requestIterator = m_requestList.begin(); requestIterator != m_requestList.end(); ++requestIterator)
	{
		if (((*requestIterator) != NULL) && ((*requestIterator)->getRequestHandle() == requestHandle))
		{
			(*aRequest) = (*requestIterator);
			ret = S_OK;
			break;
		}
	}
	m_requestListJanitor.unlock();
	return ret;
}   //  end requestCompletedInternally

//-----------------------------------------------------------------------------
// ValuesChanged
//
long DaTransaction::valuesChanged(void)
{
	m_requestListJanitor.lock();
	size_t count = m_requestList.size();

	if (count == 0)
	{
		m_requestListJanitor.unlock();
		return EnumResultCode_S_FALSE;
	}   //  end if

	OTValueData* values = new OTValueData[count];
	unsigned long* handles = new unsigned long[count];

	for (size_t i = 0; i < count; i++)
	{
		DaRequest* request = m_requestList[i];
		BOOL setOk = FALSE;

		if (request != NULL)
		{
			ValueQT* value = request->getValue();

			if (value != NULL)
			{
				values[i].m_quality = (unsigned char)value->getQuality();
				values[i].m_timestamp = value->getTimeStamp().get();
				values[i].m_pValue = new VARIANT();
				value->getData().copyTo(values[i].m_pValue);
				handles[i] = request->getAddressSpaceElement()->getObjectHandle();
				setOk = TRUE;
			}   //  end if
		}   //  end if

		if (!setOk)
		{
			values[i].m_quality     = EnumQuality_BAD;
			values[i].m_pValue      = new VARIANT();
			values[i].m_pValue->vt  = VT_EMPTY;
			handles[i] = 0;
		}   //  end if ... else

		request->setResult(EnumResultCode_S_OK);
	}   //  end for

	long result = OTSValuesChanged((long)count, handles, values);
	m_requestListJanitor.unlock();

	for (size_t j = 0; j < count; j++)
	{
		if (values[j].m_pValue != NULL)
		{
			Variant::variantClear(values[j].m_pValue);
			delete values[j].m_pValue;
			values[j].m_pValue = NULL;
		}   //  end if
	}   //  end for

	//  release the allocated memory
	if (values != NULL)
	{
		delete[] values;
	}   //  end if

	if (handles != NULL)
	{
		delete [] handles;
	}   //  end if

	return result;
}   //  end ValuesChanged


//----------------------------------------------------------------------------
// removeRequest
//
void DaTransaction::removeRequest(DaRequest* aRequest)
{
	if (aRequest == NULL)
	{
		return;
	}   //  end if

	m_requestListJanitor.lock();

	std::vector<DaRequest*>::iterator requestIterator;
	for (requestIterator = m_requestList.begin(); requestIterator != m_requestList.end(); requestIterator++)
	{
		if (*requestIterator == aRequest)
		{
			delete *requestIterator;
			m_requestList.erase(requestIterator);
			break;
		}   //  end if
	}   //  end if

	m_requestListJanitor.unlock();
}   //  end removeRequest

